src/org/sosy_lab/cpachecker/cpa/smg/evaluator/AssumeVisitor.java

```diff
@@ -59,24 +66,51 @@ public AssumeVisitor(SMGExpressionEvaluator pSmgExpressionEvaluator, CFAEdge pEd
         for (SMGValueAndState leftSideValAndState :
             smgExpressionEvaluator.evaluateExpressionValue(
                 getInitialSmgState(), edge, leftSideExpression)) {
-        SMGSymbolicValue leftSideVal = leftSideValAndState.getObject();
+          SMGValue leftSideVal = leftSideValAndState.getObject();
         SMGState newState = leftSideValAndState.getSmgState();
 
           for (SMGValueAndState rightSideValAndState :
               smgExpressionEvaluator.evaluateExpressionValue(newState, edge, rightSideExpression)) {
-          SMGSymbolicValue rightSideVal = rightSideValAndState.getObject();
+            SMGValue rightSideVal = rightSideValAndState.getObject();
           newState = rightSideValAndState.getSmgState();
 
             for (SMGValueAndState resultValueAndState :
                 evaluateBinaryAssumption(newState, binaryOperator, leftSideVal, rightSideVal)) {
               newState = resultValueAndState.getSmgState();
-              SMGSymbolicValue resultValue = resultValueAndState.getObject();
-
-              //TODO: separate modifiable and unmodifiable visitor
-              int leftSideTypeSize = smgExpressionEvaluator.getBitSizeof(edge, leftSideExpression.getExpressionType(), newState);
-              int rightSideTypeSize = smgExpressionEvaluator.getBitSizeof(edge, rightSideExpression.getExpressionType(), newState);
-              newState.addPredicateRelation(leftSideVal, leftSideTypeSize,
-                  rightSideVal, rightSideTypeSize, binaryOperator, edge);
+              SMGValue resultValue = resultValueAndState.getObject();
+
+              // TODO: separate modifiable and unmodifiable visitor
+              CType leftSideType = leftSideExpression.getExpressionType();
+              SMGType leftSideSMGType =
+                  SMGType.constructSMGType(leftSideType, newState, edge, smgExpressionEvaluator);
+              if (leftSideExpression instanceof CCastExpression) {
```

Comment:
Do you handle recursively nested casts here?
I do not know whether the CFA contains something like `(char)((int)((double)x)))`.
Please test and if necessary, replace the `if` with `while`.